components/proxy/src/main/kotlin/com/hotels/styx/StyxConfigValidation.kt

```diff
@@ -0,0 +1,54 @@
+/*
+  Copyright (C) 2013-2019 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx
+
+import com.hotels.styx.ServerConfigSchema.validateServerConfiguration
+import com.hotels.styx.config.schema.SchemaValidationException
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory.getLogger
+
+val LOG: Logger = getLogger("com.hotels.styx")
+
+/**
+ * Used to perform validation on StyxConfig.
+ * StyxConfig can only be validated if using YamlConfiguration underneath.
+ */
+// TODO it would be nice if we could implement this without violating encapsulation by knowing whether YAML is used internally
+// TODO but that should be designed carefully so that our class model still makes sense.
+fun validate(config: StyxConfig) {
```

Comment:
Here is how StyxServer is built programmatically:

```
    val styxServer = StyxServer(
            StyxServerComponents.Builder()
                    .styxConfig(StyxConfig.fromYaml(yamlText))
                    .build())
```

A `StyxConfig` object is created, and that injected to `StyxServerComponents`. As you say @dvlato it seems sensible to validate the server configuration right there in the `StyxConfig` class.

Reply:
I totally agree it seems the StyxConfig construction itself could take care of the config validation.  That might help us unify the two different 'validation points'.

Reply:
There is a down side here, though. Currently, is is possible to enable or disable config validation programmatically in `StyxServerComponents`. If we move it into `StyxConfig` we will need to either add more overloaded constructors for `disableConfigValidation` or have it always check the system property, even in tests, which could lead to tests affecting each other, since system properties live for the lifetime of the JVM.

Reply:
>There is a down side here, though. Currently, is is possible to enable or disable config validation programmatically in StyxServerComponents. If we move it into StyxConfig we 

Not on `master` though. 

> will need to either add more overloaded constructors for disableConfigValidation or 

As a starting point, try to remove `StartupConfig` out of `StyxConfig`. It is not really necessary there, but this will bring down the number of constructors. AFAICS the `StartupConfig` is necessary for the admin interface. Just pass it there separately.

Also consider removing either `StyxConfig` default constructor, or `defaultConfig` static factory method. Do we really need both?

Similarly `StyxConfig(String yaml)` should be removed in favour of `fromYaml` static factory method. You can then perform the config validation in this method before constructing the object. You can also provide an overload offering an option to disable validation:

```
public static StyxConfig fromYaml(String yamlText) // always validates
public static StyxConfig fromYaml(String yamlText, boolean validate)
```

Obviously there is no need to do anything about `StyxConfig(Configuration)` constructor because we cannot apply validation to `Configuration` objects.